lib/spack/spack/solver/asp.py

```diff
@@ -942,54 +945,62 @@ def conflict_rules(self, pkg):
                 self.gen.fact(fn.conflict(pkg.name, trigger_id, constraint_id, conflict_msg))
                 self.gen.newline()
 
-    def available_compilers(self):
+    def compiler_facts(self):
         """Facts about available compilers."""
 
-        self.gen.h2("Available compilers")
-        compilers = self.possible_compilers
+        # Build a dictionary mapping a compiler spec to supported OS and targets
+        for entry in spack.compilers.all_compilers_config():
+            compiler_entry = entry["compiler"]
+            key = spack.spec.CompilerSpec(compiler_entry["spec"])
 
-        compiler_versions = collections.defaultdict(lambda: set())
-        for compiler in compilers:
-            compiler_versions[compiler.name].add(compiler.version)
+            # FIXME: same spec with multiple OS and targets?
+            if key in self.compiler_info:
+                warnings.warn(
+                    f"compiler {key} has duplicate entries, only the first one will be considered"
+                )
```

Comment:
> I think this should just take the first one without warning. 

I added the warning because right now talking the first one doesn't mean we take the _right_ one (the first entry for compiler spec `X@Y` might be using the wrong OS, or the wrong target). For instance with a configuration like:
```yaml
compilers:
- compiler:
    spec: gcc@9.4.0
    paths:
      cc: /usr/bin/gcc
      cxx: /usr/bin/g++
      f77: /usr/bin/gfortran
      fc: /usr/bin/gfortran
    operating_system: ubuntu20.04
    target: x86_64
- compiler:
    spec: gcc@9.4.0
    paths:
      cc: /usr/bin/gcc
      cxx: /usr/bin/g++
      f77: /usr/bin/gfortran
      fc: /usr/bin/gfortran
    operating_system: ubuntu20.04
    target: aarch64
```
the first entry is wrong if we are targeting an `aarch64`. 

---

Regarding flags, the new encoding is a step forward towards differentiating compilers, since the encoding for this:
```yaml
compilers:
- compiler:
    spec: gcc@9.4.0
    paths:
      cc: /usr/bin/gcc
      cxx: /usr/bin/g++
      f77: /usr/bin/gfortran
      fc: /usr/bin/gfortran
    flags:
      cflags: "-O2"
    operating_system: ubuntu20.04
    target: x86_64
```
is the following:
```
compiler_id(2).
compiler_name(2,"gcc").
compiler_version(2,"9.4.0").
compiler_os(2,"ubuntu20.04").
compiler_target(2,"x86_64").
compiler_flag(2,"cflags","-O2").
```
This means that the new encoding in ASP can differentiate two compiler with the same name and version but different flags. The warning is there since we have no way at the moment to select them from the spec syntax[^1].

[^1]: The workaround for this use case, as far as I know is to tweak the version when another unselectable parameter is different.

Reply:
The warning is now triggered if we have compilers which can't be distinguished using the spec syntax (same spec, os and target).